R/fill_ch_names.R

```diff
@@ -43,187 +58,721 @@ fill_ch_names <- function(ch_data,
+      .data[["match_distance_jaccard"]]
+    ) %>%
+    dplyr::distinct(.data[["unique_identifier"]], .keep_all = TRUE) %>%
+    dplyr::select(dplyr::all_of(col_to_select))
 
-  no_match_pc_missing <- ch_data %>%
-    dplyr::anti_join(ch_name_lookup,
-      by = dplyr::join_by("ch_postcode"),
-      na_matches = "never"
+  ch_match <- dplyr::bind_rows(
+    ch_pc_match,
+    ch_name_match1,
+    ch_name_match2,
+    ch_name_match3
+  )
+
+  ### ch_postcode and postcode exchange, then matching, quality 18L----
+  ch_pc_exchange_match1 <- ch_data %>%
+    dplyr::anti_join(ch_match,
+      by = dplyr::join_by("unique_identifier")
     ) %>%
-    dplyr::filter(
-      !is.na(.data[["ch_name"]]) & is.na(.data[["ch_postcode"]])
+    dplyr::mutate(
+      intermediate_pc = .data[["ch_postcode"]],
+      ch_postcode = .data[["postcode"]],
+      postcode = .data[["ch_postcode"]],
+      ch_pc_partial = stringr::str_sub(.data[["ch_postcode"]], 1, -2),
+      ch_pc_partial2 = stringr::str_sub(.data[["ch_postcode"]], 1, -3),
+      ch_pc_partial3 = stringr::str_sub(.data[["ch_postcode"]], 1, -5)
     ) %>%
-    dplyr::left_join(ch_name_best_match,
+    dplyr::select(-.data[["intermediate_pc"]]) %>%
+    dplyr::inner_join(
+      ch_name_lookup,
       by = dplyr::join_by(
-        "ch_name",
-        closest("ch_admission_date" <= "latest_close_date")
+        x$ch_name_keyword == y$ch_name_validated_keyword,
+        x$ch_admission_date <= y$latest_close_date,
+        x$ch_admission_date >= y$ch_date_registered,
+        "ch_pc_partial"
       ),
-      multiple = "last",
-      na_matches = "never",
-      suffix = c("_old", "")
+      na_matches = "never"
     ) %>%
     dplyr::mutate(
-      ch_postcode = dplyr::if_else(!is_missing(.data[["ch_postcode"]]),
-        .data[["ch_postcode"]],
-        .data[["ch_postcode_old"]]
+      ch_name_old = .data[["ch_name"]],
+      ch_postcode_old = .data[["ch_postcode"]],
+      ch_name = .data[["ch_name_validated"]],
+      ch_postcode = .data[["ch_postcode_lookup"]],
+      matching_quality_indicator_overall = 18L,
+      match_distance_jaccard = stringdist::stringdist(.data[["ch_name"]],
+        .data[["ch_name_validated"]],
+        method = "jaccard"
       )
-    )
+    ) %>%
+    dplyr::arrange(
+      .data[["unique_identifier"]],
+      .data[["match_distance_jaccard"]]
+    ) %>%
+    dplyr::distinct(.data[["unique_identifier"]], .keep_all = TRUE) %>%
+    dplyr::select(dplyr::all_of(col_to_select))
 
-  no_match_name_missing <- ch_data %>%
-    dplyr::anti_join(ch_name_lookup,
-      by = dplyr::join_by("ch_postcode"),
+  ch_match <- dplyr::bind_rows(
+    ch_pc_match,
+    ch_name_match1,
+    ch_name_match2,
+    ch_name_match3,
+    ch_pc_exchange_match1
+  )
+
+  ## Other matching processes ----
+  ### quality 19L----
+  # ch_postcode and postcode exchange,
+  # then fizzy match ch_name, and matching main part of postcode
+  ch_pc_exchange_match2 <- ch_data %>%
```

Comment:
This is where I realised that there are English Care homes in the data but they are not in the Care Inspectorate spreadsheet (only Scottish care homes are in the spreadsheet). In these cases the patient will have a scottish postcode but the ch postcode will be in England. 
These happen because there are no places at scottish care homes for the patient to go (specific care needs etc) so they get a bed in an English care home but it is still funded by scotland. 

Solve for this would be to remove cases where the ch_postcode does not appear in the scottish postcode directory.

Reply:
English care homes added, quality 15.